src/cluster_import.c

```diff
@@ -0,0 +1,1915 @@
+              link->linkname,
+              link->nodename,
+              link->slot_ranges_str);
+    updateSlotMigrationLinkState(link, SLOT_IMPORT_RECONNECT);
+}
+
+void clusterHandleSlotImportLinkClientOOM(void *import) {
+    slotMigrationLink *link = (slotMigrationLink *)import;
+    serverAssert(link->type == SLOT_MIGRATION_IMPORT);
+    if (isSlotMigrationLinkInProgress(link)) {
+        serverLog(LL_WARNING,
+                  "Slot import link %.40s to node %.40s (owner of slots [%s]) "
+                  "failed due due to OOM",
+                  link->linkname,
+                  link->nodename,
+                  link->slot_ranges_str);
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED,
+                                "Ran out of memory (OOM) during slot import");
+    }
+}
+
+/* Validate that the slot migration is still owned by the node it originally
+ * was importing from. If the source node has since changed, or no longer has
+ * sole ownership by the original source node, we fail the migration. Otherwise,
+ * return the source node for this migration. */
+clusterNode *validateClusterNodeForImportIsUnchanged(slotMigrationLink *link) {
+    serverAssert(link->type == SLOT_MIGRATION_IMPORT);
+    int cross_node;
+    clusterNode *n = getClusterNodeBySlotRanges(link->slot_ranges, &cross_node);
+    if (n && !memcmp(n->name, link->nodename, CLUSTER_NAMELEN)) {
+        return n;
+    }
+    if (n == server.cluster->myself) {
+        serverLog(LL_WARNING,
+                  "Slot import link %.40s to node %.40s (previous owner of slots [%s]) "
+                  "already owned by myself. Failing slot import...",
+                  link->linkname,
+                  link->nodename,
+                  link->slot_ranges_str);
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED,
+                                "Slots were unexpectedly assigned to myself during import");
+        return NULL;
+    }
+    serverLog(LL_WARNING,
+              "Slot import link %.40s to node %.40s (previous owner of slots [%s]) "
+              "contains at least one slot no longer owned by the source node. "
+              "Failing slot import...",
+              link->linkname,
+              link->nodename,
+              link->slot_ranges_str);
+    finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED,
+                            "Slots are no longer owned by source node");
+    return NULL;
+}
+
+void performSlotImportLinkReconnect(slotMigrationLink *link) {
+    serverAssert(link->type == SLOT_MIGRATION_IMPORT);
+    /* Leaving partially imported slots would result in corruption. */
+    resetSlotMigrationLink(link);
+    delKeysNotOwnedByMyself(link->slot_ranges);
+
+    clusterNode *n = validateClusterNodeForImportIsUnchanged(link);
+
+    if (n) {
+        /* Do the reconnect for this link. */
+        connectSlotImportLink(link, n->ip, getNodeDefaultClientPort(n));
```

Comment:
Calling these links is a bit confusing, since today links refer to the clusterbus links, but this is basically just a normal client.

Reply:
Hmm. Yeah, your point makes sense. I guess I would address the comment to be "Do the reconnect for this nodes cluster import link then". I think just saying this (node's) link will be confusing. I think we can use that terminology throughput.

Reply:
Makes sense. This comment is gone, but I have tried to apply "slot migration link" nomenclature this throughout